File: 351001/Betenya/backend/src/auth/auth.service.ts
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { PrismaService } from '../services/prisma.service';
import { JwtService } from './jwt.service';
import { RegisterRequestTo } from '../dto/auth/RegisterRequestTo.dto';
import { LoginRequestTo } from '../dto/auth/LoginRequestTo.dto';
import * as bcrypt from 'bcrypt';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  async register(dto: RegisterRequestTo) {
    const existing = await this.prisma.user.findUnique({
      where: { login: dto.login },
    });

    if (existing) {
      throw new HttpException(
        { errorMessage: 'User with this login already exists', errorCode: 40300 },
        HttpStatus.FORBIDDEN,
      );
    }

    const salt = await bcrypt.genSalt();
    const hashedPassword = await bcrypt.hash(dto.password, salt);

    const user = await this.prisma.user.create({
      data: {
        login: dto.login,
        password: hashedPassword,
        firstname: dto.firstname,
        lastname: dto.lastname,
        role: dto.role || 'CUSTOMER',
      },
    });

    return {
      id: user.id,
      login: user.login,
      firstname: user.firstname,
      lastname: user.lastname,
      role: user.role,
    };
  }

  async login(dto: LoginRequestTo) {
    const user = await this.prisma.user.findUnique({
      where: { login: dto.login },
    });

    if (!user) {
      throw new HttpException(
        { errorMessage: 'Invalid credentials', errorCode: 40100 },
        HttpStatus.UNAUTHORIZED,
      );
    }

    const passwordMatch = await bcrypt.compare(dto.password, user.password);
    if (!passwordMatch) {
      throw new HttpException(
        { errorMessage: 'Invalid credentials', errorCode: 40100 },
        HttpStatus.UNAUTHORIZED,
      );
    }

    const token = this.jwtService.sign({
      sub: user.login,
      id: Number(user.id),
      role: user.role,
    });

    return { access_token: token };
  }
}

File: 351004/Kostukov/publisher/app/services/auth_service.py
from fastapi import HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from publisher.app.infrastructure.db.repo import WriterRepo
from publisher.app.core.security.password import hash_password, verify_password
from publisher.app.core.security.jwt import create_access_token


class AuthService:
    def __init__(self):
        self.writer_repo = WriterRepo()

    async def register_writer(self, session: AsyncSession, dto):
        existing = await self.writer_repo.get_by_login(session, dto.login)
        if existing:
            raise HTTPException(status_code=409, detail="Login already exists")

        data = {
            "login": dto.login,
            "password": hash_password(dto.password),
            "firstname": dto.first_name,
            "lastname": dto.last_name,
            "role": dto.role,
        }

        return await self.writer_repo.create(session, data)

    async def login(self, session: AsyncSession, dto):
        writer = await self.writer_repo.get_by_login(session, dto.login)
        if not writer:
            raise HTTPException(status_code=401, detail="Invalid login or password")

        if not verify_password(dto.password, writer.password):
            raise HTTPException(status_code=401, detail="Invalid login or password")

        token = create_access_token(login=writer.login, role=writer.role)

        return {
            "access_token": token,
            "token_type": "Bearer",
        }